src/pybind/mgr/cephadm/serve.py

```diff
@@ -583,30 +575,41 @@ def _apply_all_services(self) -> bool:
         specs = []  # type: List[ServiceSpec]
         for sn, spec in self.mgr.spec_store.active_specs.items():
             specs.append(spec)
+        apply_spec_fails = []
         for spec in specs:
             try:
                 if self._apply_service(spec):
                     r = True
             except Exception as e:
-                self.log.exception('Failed to apply %s spec %s: %s' % (
-                    spec.service_name(), spec, e))
+                msg = f'Failed to apply {spec.service_name()} spec {spec}: {str(e)}'
+                self.log.exception(msg)
                 self.mgr.events.for_service(spec, 'ERROR', 'Failed to apply: ' + str(e))
+                warning_name = f'{spec.service_name()}_APPLY_SPEC_FAIL'
```

Comment:
The idea is that the warning name is a fixed string that is well-documented here: https://docs.ceph.com/en/latest/rados/operations/health-checks/

In this case, I think we want ``CEPHADM_APPLY_SPEC_FAIL`` or similar.  Since we need multiple detail messages/error, potentially 1 or more per spec, I would suggest detail strings like ``nfs.foo: {e}``.  This is a bit more awkward because this function will probably need to (1) build a list of error message(s) for this service, (2) examine the existing list of detail strings for CEPHADM_APPLY_SPEC_FAIL based on matching by string prefix ``{spec.name}: ``, and (3) update the checks.  The set/remove helpers you added won't really help you here; I think you'll need a new one that does the prefix matching and handles the case where pass an empty list of errors, clears the alert when there are no remaining items (across any service), etc... does that make sense?

Reply:
Is it possible we could work this like the CEPHADM_STRAY_DAEMON (or stray host) check? That check clears itself every time right before it it goes to check for stray daemons/hosts again and then sets the health warning again if it finds stray daemons/hosts again https://github.com/ceph/ceph/blob/master/src/pybind/mgr/cephadm/serve.py#L515-L579. Since we apply every spec each time it could possibly work in the same way: we just build up the list of errors as we apply all the services, delete the old value for the check, then set the check value to whatever we found this time around (if we have anything). It would save the trouble of having to look through the old list and have to match prefixes and delete only certain portions. Unless we feel we need to keep errors that happened on previous iterations of the serve loop that didn't happen this time around I feel like it could work.

Reply:
@liewegas @adk3798 ty for the ideas! I see what you're both saying and I'm going to change it. If we go with Adam's suggestion on clearing the warning, I think that makes clearing pretty easy. 

But the thing I'm wondering now is how to set the health warning? Since Sage said that we want the warning name to be a fixed string, if there were multiple warnings with the same name, they would just overwrite each other because they all have the same name/key in the `self.mgr.health_checks` dict. In the health checks, the `summary` string is what's shown when you do `ceph -s` or `ceph health`, so my idea right now is to build up the list of errors, and concatenate the `summary` string with `;` between each error string within their respective error, then set the health warning at the end.

Reply:
@melissa-kun-li I think we could follow the stray daemon/host way of doing things for that as well. The summary string itself could just say something like "x services failed to be applied" where x is how many specs through errors getting applied. The actual list of errors that caused those failures could be put in the details section of the health check which can then be seen by the user if they run `ceph health detail`. That way we have a concise message in `ceph -s` to notify users their is an issue and they can check the details of that error with `ceph health detail`.

Reply:
oooh I see, that sounds good I'll try that!